### PR TITLE
TFNetworkLayer.py added flag masked_time to BatchNorm layer

### PR DESCRIPTION
**Pull request by squarenabla**
When seq_lens are equal, possible faster computation with masked_time=False.
Some modifications in TFUtil.py, which fix bugs and extend the functionality.
See #88 for previous conversations.

### CODE REVIEW

File: TFNetworkLayer.py [hunk 1]
```diff
@@ -895,6 +897,8 @@ def batch_norm(self, data,
     :param tf.Tensor sample_variance:
     :param tf.Tensor gamma:
     :param tf.Tensor beta:
+    :param bool masked_time: flatten and mask input tensor
+    :param tf.DType dtype: data type, different from default when testing
```
Comment (albertz) [inline TFNetworkLayer.py]: `dtype` can also be of type `str`, right? Or probably that is even the usual case which you often have. You should write that, i.e. `:param tf.DType|str dtype:`.

But anyway, I don't quite understand why you need that? It should not have an effect. Also, it's not clear from the documentation and also from the parameter name what this is really used for. Please remove this if this is not really needed.

File: TFNetworkLayer.py [hunk 2]
```diff
@@ -904,8 +908,12 @@ def batch_norm(self, data,
...
     """
     with tf.variable_scope("batch_norm"):
-      x = data.get_placeholder_flattened(keep_dims=True)  # shape (time',...)
-      mean, variance = tf.nn.moments(x, axes=[0], keep_dims=True)
+      if masked_time:
+        x = data.get_placeholder_flattened(keep_dims=True)
+        mean, variance = tf.nn.moments(x, axes=[0], keep_dims=True)
+      else:
+        x = data.placeholder
+        mean, variance = tf.nn.moments(x, axes=data.get_axes(exclude_feature=True), keep_dims=True)
```
Comment (albertz) [inline TFNetworkLayer.py]: This is not consistent, and thus wrong.
E.g. consider you have `data` with (batch) shape (B, T, H, D).
In the `masked_time` case, you calculate mean/var over (B, T), i.e. the mean/var shape is (1, 1, H, D).
In the other case, you calculate mean/var over (B,T,H), i.e. the mean/var shape is (1,1,1,D).

Comment (albertz) [reply TFNetworkLayer.py]: The question is also, what do you really want to have here? If you need that new behavior (mean/var over (B,T,H)), then that should be yet another option.
Btw, in both cases (`masked_time` or not), you can use the same `axes` argument, because it should be the same.

File: TFNetworkLayer.py [hunk 3]
```diff
@@ -924,7 +932,7 @@ def batch_norm(self, data,
             trainable=False))
         sample_variance = tf.assign_add(sample_variance, (variance - sample_variance) * momentum)
       # If train or if force_sample, use default use_sample=0.0, otherwise use_sample=1.0.
-      use_sample = 1.0 + tf.cast(tf.logical_or(self.network.train_flag, force_sample), tf.float32) * (use_sample - 1.0)
+      use_sample = 1.0 + tf.cast(tf.logical_or(self.network.train_flag, force_sample), dtype) * (use_sample - 1.0)
```
Comment (albertz) [inline TFNetworkLayer.py]: Why?

File: TFUtil.py [hunk 4]
```diff
@@ -1028,10 +1028,11 @@ def get_placeholder_flattened(self, keep_dims=False):
       ndim -= 1
     if len(dyn_axes) > 1:
       shape = tf.shape(x)
+      shape_as_list = x._shape_as_list()
```
Comment (albertz) [inline TFUtil.py]: You are calling a hidden method. You should not do that. Also, why do you need that?

File: TFUtil.py [hunk 5]
```diff
@@ -1028,10 +1028,11 @@ def get_placeholder_flattened(self, keep_dims=False):
...
       x = tf.reshape(
         x,
         [tf.reduce_prod([shape[i] for i in dyn_axes])] +
-        [shape[i] for i in range(ndim) if i not in dyn_axes])
+        [shape_as_list[i] for i in range(ndim) if i not in dyn_axes])
```
Comment (albertz) [inline TFUtil.py]: Why? The old could was just the same.

File: tests/test_TFNetworkLayer.py [hunk 6]
```diff
@@ -146,6 +146,134 @@ def test_batch_norm_vars():
...
+    net.train_flag = True
+    with tf.variable_scope("src_nhwc"):
+      src_nhwc = InternalLayer(name="src_nhwc", network=net, out_type={"dim": 16,
+                                                                       "shape": (None, 16, 16),
+                                                                       "batch_dim_axis": 0,
+                                                                       "time_dim_axis": 1,
+                                                                       "feature_dim_axis": 3,
+                                                                       "sparse": False
+                                                                       })
+      src_nhwc.output.placeholder = tf.placeholder(shape=(None, None, 16, 16), dtype=tf.float64)
```
Comment (albertz) [inline tests/test_TFNetworkLayer.py]: The `dtype` is wrong. It should be `tf.float32`.

File: tests/test_TFNetworkLayer.py [hunk 7]
```diff
@@ -146,6 +146,134 @@ def test_batch_norm_vars():
...
+    assert np.array_equal(np_bn2.astype('f'), out_2.astype('f'))
+    # Manually calculating batch_norm with different seq_lens, having:
+    # Mean_1 = n2 / n1 * Mean_2
+    # Var_1 = n2 / n1 * (Var_2 + Mean_2 ^ 2 (1 - n2 / n1))
+    # bn_1 = (x - Mean_1) * 1 / sqrt(Var_1)
+    # Substituting Mean_1 and Var_1:
+    np_bn1 = (input_data - n2 / n1 * data_mean) * \
+             (1.0 / np.sqrt(n2 / n1 * (data_var + data_mean ** 2 * (1 - n2 / n1))))
+    # Check with tf output.
+    assert np.array_equal(np_bn1.astype('f'), out_1.astype('f'))
```
Comment (albertz) [inline tests/test_TFNetworkLayer.py]: I think you rather should use sth like `np.testing.assert_almost_equal` or so.

File: TFNetworkLayer.py (hunk 1 above)
Comment (squarenabla) [reply TFNetworkLayer.py]: For some reason numpy and tf calculate mean differently (diff is about e-4 or e-5). The calculations are more accurate with float64 type. So, I need that to do tests and assert on equal results.

Comment (albertz) [reply TFNetworkLayer.py]: No, you do not need that. Do not assert on equal results anyway. You should use `numpy.testing.assert_almost_equal`.

File: TFNetworkLayer.py (hunk 2 above)
Comment (albertz) [reply TFNetworkLayer.py]: Also, I wonder, why did you not catch this error in your test case? That means that your test case is likely broken.

File: tests/test_TFNetworkLayer.py [hunk 8]
```diff
@@ -146,6 +146,134 @@ def test_batch_norm_vars():
...
+                                                                       "batch_dim_axis": 0,
+                                                                       "time_dim_axis": 1,
+                                                                       "feature_dim_axis": 3,
+                                                                       "sparse": False
+                                                                       })
+      src_nhwc.output.placeholder = tf.placeholder(shape=(None, None, 16, 16), dtype=tf.float32)
+      src_nhwc.output.size_placeholder = {0: tf.placeholder(shape=(None,), dtype=tf.int32)}
+
+    mean =  tf.constant(np.random.rand(1, 1, 16, 16), name="rand_mean", dtype=tf.float32)
+    variance = tf.constant(np.random.rand(1, 1, 16, 16), name="rand_var", dtype=tf.float32)
```
Comment (albertz) [inline tests/test_TFNetworkLayer.py]: You should use `rnd = numpy.random.RandomState(42)` or so initially, such that you get deterministic behavior.

File: tests/test_TFNetworkLayer.py [hunk 9]
```diff
@@ -146,6 +146,134 @@ def test_batch_norm_vars():
...
+                                                                       "shape": (None, 16, 16),
+                                                                       "batch_dim_axis": 0,
+                                                                       "time_dim_axis": 1,
+                                                                       "feature_dim_axis": 3,
+                                                                       "sparse": False
+                                                                       })
+      src_nhwc.output.placeholder = tf.placeholder(shape=(None, None, 16, 16), dtype=tf.float64)
+      src_nhwc.output.size_placeholder = {0: tf.placeholder(shape=(None,), dtype=tf.int64)}
+    mean = tf.constant(np.random.rand(1, 1, 1, 16), name="rand_mean", dtype=tf.float64)
+    variance = tf.constant(np.random.rand(1, 1, 1, 16), name="rand_var", dtype=tf.float64)
```
Comment (albertz) [inline tests/test_TFNetworkLayer.py]: This is wrong. It should be of shape (1, 1, 16, 16).

File: tests/test_TFNetworkLayer.py [hunk 10]
```diff
@@ -146,6 +146,134 @@ def test_batch_norm_vars():
...
+    input_data = np.random.rand(10, 11, 16, 16)
+    input_data[2, 5:, :, :] = 0
+    data_mean = np.mean(input_data, axis=(0, 1, 2), keepdims=True, dtype=np.float64)
+    data_var = np.var(input_data, axis=(0, 1, 2), keepdims=True, dtype=np.float64)
```
Comment (albertz) [inline tests/test_TFNetworkLayer.py]: This is wrong. The axis to reduce are only (0,1). But see my comment above also on that you have that inconsistent with `masked_time`.

File: TFNetworkLayer.py (hunk 1 above)
Comment (albertz) [reply TFNetworkLayer.py]: No, you do not need that. Use `numpy.testing.assert_almost_equal`.

File: tests/test_TFNetworkLayer.py (hunk 9 above)
Comment (squarenabla) [reply tests/test_TFNetworkLayer.py]: No, I don't use batch_norm in 2 dims. Just in the last dim. So, the statistics is gathered over 0, 1 and 2 axes. I'll make another pull request with multiple axes support. There're a few issues, which have to be tested separately. E.g. saving the bn weights in the NCHW and restoring in the NHWC mode (I told you about that problem).

File: TFNetworkLayer.py (hunk 2 above)
Comment (squarenabla) [reply TFNetworkLayer.py]: > In the `masked_time` case, you calculate mean/var over (B, T), i.e. the mean/var shape is (1, 1, H, D).
No, the mean/var shape is (1, 1, 1, D), which is a default behavior. `data.get_placeholder_flattened(keep_dims=True)` returns (BxTxH, D) tensor.

File: TFUtil.py (hunk 5 above)
Comment (squarenabla) [reply TFUtil.py]: Technically it was the same, but `[shape[i] for i in range(ndim) if i not in dyn_axes]` consists of tensors and returns None shape for static dims. However it's possible to specify the static dims by gettign the values from `x.get_shape().as_list()`.

File: TFNetworkLayer.py (hunk 2 above)
Comment (albertz) [reply TFNetworkLayer.py]: Oh yes, you are right.

File: TFUtil.py (hunk 5 above)
Comment (albertz) [reply TFUtil.py]: No, `shape[i]` is never `None`. But `shape_as_list[i]` could be `None`, so this would be wrong.

File: tests/test_TFNetworkLayer.py [hunk 11]
```diff
@@ -146,6 +146,131 @@ def test_batch_norm_vars():
...
+                                                                       "time_dim_axis": 1,
+                                                                       "feature_dim_axis": 3,
+                                                                       "sparse": False
+                                                                       })
+      src_nhwc.output.placeholder = tf.placeholder(shape=(None, None, 16, 16), dtype=tf.float32)
+      src_nhwc.output.size_placeholder = {0: tf.placeholder(shape=(None,), dtype=tf.int32)}
+
+    rnd = np.random.RandomState(42)
+    mean =  tf.constant(rnd.rand(1, 1, 16, 16), name="rand_mean", dtype=tf.float32)
+    variance = tf.constant(rnd.rand(1, 1, 16, 16), name="rand_var", dtype=tf.float32)
```
Comment (albertz) [inline tests/test_TFNetworkLayer.py]: Sorry, I was wrong before. This should be the same shape as it is in the `batch_norm` function.

File: TFUtil.py (hunk 5 above)
Comment (squarenabla) [reply TFUtil.py]: `shape[i]` is not `None` but it's a tensor. After `tf.reshape` with a tensor, the resulting dim becomes None even it was defined before.

Comment (albertz) [reply TFUtil.py]: Yes, why does that matter? This is the argument for `tf.reshape`, and it must not be `None`.

File: tests/test_TFNetworkLayer.py (hunk 11 above)
Comment (squarenabla) [reply tests/test_TFNetworkLayer.py]: Yes, didn't notice. The shapes here should also be `(1, 1, 1, 16)`.

File: TFUtil.py (hunk 5 above)
Comment (squarenabla) [reply TFUtil.py]: The result of `tf.reshape` with Tensor as an argument gives `None`. Even for `dim`.

**Comment by squarenabla** [comment]
For some reason, the tests fail with python2.7. However everything is ok with python3.6. I'll fix the problem.

File: tests/test_TFNetworkLayer.py [hunk 12]
```diff
@@ -146,6 +146,131 @@ def test_batch_norm_vars():
...
+    # Manually calculating batch_norm and compare to the tf output
+    np_bn2 = (input_data - data_mean) * (1.0 / np.sqrt(data_var))
+    npt.assert_array_almost_equal(np_bn2, out_2, decimal=5)
+    # Manually calculating batch_norm with different seq_lens, having:
+    # Mean_1 = n2 / n1 * Mean_2
+    # Var_1 = n2 / n1 * (Var_2 + Mean_2 ^ 2 (1 - n2 / n1))
+    # bn_1 = (x - Mean_1) * 1 / sqrt(Var_1)
+    # Substituting Mean_1 and Var_1:
+    np_bn1 = (input_data - n2 / n1 * data_mean) * \
+             (1.0 / np.sqrt(n2 / n1 * (data_var + data_mean ** 2 * (1 - n2 / n1))))
```
Comment (albertz) [inline tests/test_TFNetworkLayer.py]: `n2`/`n1` is an `int`, so this behaves different in Pyhton 2/3.
You maybe should add `from __future__ import division` in the beginning of the file, then they are always the same.

File: TFUtil.py (hunk 5 above)
Comment (albertz) [reply TFUtil.py]: This is not the result of `tf.reshape`, this is the argument to `tf.reshape`. For that argument, you must not give `None`.

Comment (albertz) [reply TFUtil.py]: Maybe you misunderstand me. To make it clear:
Sth like this is allowed:

    tf.reshape(x, (1, 2, 3, some_tensor, 4, 5))

Sth like this is **not** allowed (i.e. `None` is not allowed):

    tf.reshape(x, (1, 2, None, some_tensor, 4, 5))

Comment (squarenabla) [reply TFUtil.py]: I know, that this is not a result of `tf.reshape`. My point is: why should we pass a tensor as an argument (smth like `tf.reshape(x, (1, 2, 3, some_tensor, 4, 5))`), when we could explicitly pass a number there (`tf.reshape(x, (1, 2, 3, 34, 4, 5))`))?

Comment (albertz) [reply TFUtil.py]: `shape_as_list[i]` can be `None`. You must not pass `None`.

Comment (squarenabla) [reply TFUtil.py]: Yes, it can be. But `[shape_as_list[i] for i in range(ndim) if i not in dyn_axes]` doesn't have `None`. because these are such elements, which are `not in dyn_axes`.

Comment (albertz) [reply TFUtil.py]: Do not rely on that. That does not always have to be the case. `shape_as_list[i]` could be `None`. E.g. the feature dim can be None, and is not part of `dyn_axes` here. But there could be other cases. Even if this works now, for some of the cases, this might break in the future. Don't write code like that.
